# ChatGPT Logs
Questions 1 to 7: https://chatgpt.com/share/5decf5c5-58c6-475f-869d-a827d18f0f1e

Question 8: https://chatgpt.com/share/66e358b7-1088-8011-a5f1-0b4ca5b2b25f

In [ ]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

In [ ]:
rows, columns = df.shape
print(f"The DataFrame has {rows} rows and {columns} columns.")

The DataFrame has 391 rows and 11 columns.


# Observations and Variables
1. Observations are the individual datapoints that exist in our dataset. In the context of this dataset, that would be refering to each individual NPC in Animal Crossing. Each row of our dataset contains information on one NPC or one observation.
2. Variables are the many attributes of these NPCs that are stored within each row. Each column is a different attribute or a different variable.

In [ ]:
df['species'].describe()

count     391
unique     35
top       cat
freq       23
Name: species, dtype: object

In [ ]:
df['gender'].value_counts()

gender
male      204
female    187
Name: count, dtype: int64

In [ ]:
url2 = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df2 = pd.read_csv(url2)
df2.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

In [ ]:
df2.shape

(891, 15)

# Discrepancies Between Describe and Shape
When using the shape attribute of the DataFrame, we see that it has 891 rows and 15 columns but when using the describe method, we see that some of the columns have less than 891 entries. This is because the describe method does not count the non-numeric or the missing datapoints.

# Difference Between Attributes and Methods
- Attributes are properties that belong to an object. shape is an attribute because it belongs to the DataFrame object called df2.
- Methods are functions that are associated with the object. describe() is a method because it calls a functions belonging to the DataFrame object which then calculates things sich as the mean, standard deviation, max/min, etc.

# DataFrame.describe() Variable Meanings
- Count: The number of non-null entries in this column.
- Mean: The sum of the column divided by the number of entries (the average entry).
- std: The standard deviation. A way to measure how far the datapoints lie from their trendline.
- min: The smallest value in the column.
- 25%: The first quartile. 25% of the data falls below this value.
- 50%: The median (aka the second quartile). 50% of data falls below this value.
- 75%: The third quartile. 75% of data falls below this value.
- max: The largest value in the column.

# `df.dropna()` vs `del df['col']`
df.dropna() will drop all rows that contain missing data from the dataframe. whereas del df['col'] will completely remove that column from all rows.


1. **Provide an example of a "use case" in which using df.dropna() might be peferred over using del df['col'].**

* For our initial dataset with the Animal Crossing NPCs, if we want to perform some analysis on the NPCs songs, we can use `df.dropna()`. Some NPCs in the dataframe do not have an entry for 'song'. We can use dropna() to remove those NPCs from our database and still continue with our analysis of the remaining NPCs songs. Using del df['col'] would remove all the NPCs songs from our dataframe, making it impossible for us to conduct the analysis.

2. **Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna()**

* In our Animal Crossing NPC dataset, there is a 'url' column which contains a url to an image of the character. Suppose we do not need to perform any alaysis on the appearance of the character. In this senario, we can use `del df['url']` to remove this column from our dataframe. This could be useful if we had an extremely large dataset that is taking up an unessesary amount of space.

3. **Discuss why applying del df['col'] before df.dropna() when both are used together could be important**

* When using both del df['col'] and df.dropna(), we must apply del df['col'] first because if a row had missing data in the column that we are deleting, the row is still useful because we do not care about what is in that column. If an NPC does not have an entry in their 'url' column but they have everything, we can still use that npc to perform our analysis on the songs. If we use dropna(), it will remove them from the dataset but if we delete the column first, then there will not be any null datapoints for dropna() to detect.

4. **Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.**

* I will use del df['id'] to remove the id column because the id column contains info on the names of the NPCs in the game files which we do not need. This will remove one missing datapoint. Next, I will use df.dropna() to remove the rows which have missing datapoints. Since the only other datapoints that are missing are the songs, this will remove any NPC that is not associated with a song.

In [ ]:
if 'id' in df.columns:  # I am including this because if I run this cell more than once, it is causing an error
    del df['id']
df = df.dropna()
df.isna().sum()

row_n          0
name           0
gender         0
species        0
birthday       0
personality    0
song           0
phrase         0
full_id        0
url            0
dtype: int64

In [ ]:
df2.groupby("sex")["survived"].describe()

count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

In [ ]:
df2.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

# Why do they show different count values (`df.describe() vs df.groupby('col1')['col2'].describe()`)
df.describe() shows the count based on the entire dataframe. It shows a different number for each column depending on how many non-missing datapoints there are in that specific column. We can see that most of the columns have a count of 891 because there were 891 passengers but there were a few whos age, deck, and embark town were unown. Because of this, those columns show a number less than 891.

df.groupby("sex")["survived"].describe() shows the count of non-missing values in each group (male or female). Since there was no missing data on their sex or survivorship, this method accounts for the full 891 passengers.

  A. ChatGPT was able to immediately find the error. Using a ChatBot was easier than using google for this issue.

  B. ChatGPT was able to immediately find the error. Using a ChatBot was easier than using google for this issue.

  C. ChatGPT was able to immediately find the error. Using a ChatBot was easier than using google for this issue.

  D. ChatGPT said that the issue was that the square bracket was misplaced. The actual issue is that there was a missing round bracket. Although it said the issue was with the square bracket, the code it generated correctly replaced the round bracket so it worked fine. Using a ChatBot was easier than using google for this issue.

  E. ChatGPT was able to immediately find the error. Using a ChatBot was easier than using google for this issue.

  F. ChatGPT was able to immediately find the error. Using a ChatBot was easier than using google for this issue.

  G. ChatGPT was able to immediately find the error. Using a ChatBot was easier than using google for this issue.

  ***In conclusion, ChatBots are an amazingly powerful new tool in our arsenal which makes coding and data manipulation immensly more efficient.***